apps/emqx_machine/src/emqx_cluster_rpc.erl

```diff
@@ -76,28 +75,35 @@ start_link(Node, Name, RetryMs) ->
     TnxId :: pos_integer(),
     Reason :: string().
 multicall(M, F, A) ->
-    multicall(M, F, A, timer:minutes(2)).
+    multicall(M, F, A, all, timer:minutes(2)).
 
--spec multicall(Module, Function, Args, Timeout) -> {ok, TnxId, term()} |{error, Reason} when
+-spec multicall(Module, Function, Args, SucceedNum, Timeout) -> {ok, TnxId, term()} |{error, Reason} when
     Module :: module(),
     Function :: atom(),
     Args :: [term()],
+    SucceedNum :: pos_integer() | all,
     TnxId :: pos_integer(),
     Timeout :: timeout(),
     Reason :: string().
-multicall(M, F, A, Timeout) ->
+multicall(M, F, A, SucceedNum, Timeout) ->
     MFA = {initiate, {M, F, A}},
-    case ekka_rlog:role() of
-        core -> gen_server:call(?MODULE, MFA, Timeout);
-        replicant ->
-            %% the initiate transaction must happened on core node
-            %% make sure MFA(in the transaction) and the transaction on the same node
-            %% don't need rpc again inside transaction.
-            case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
-                {ok, Node} -> gen_server:call({?MODULE, Node}, MFA, Timeout);
-                disconnected -> {error, disconnected}
-            end
-    end.
+    Begin = erlang:monotonic_time(),
+    InitRes =
+        case ekka_rlog:role() of
+            core -> gen_server:call(?MODULE, MFA, Timeout);
+            replicant ->
+                %% the initiate transaction must happened on core node
+                %% make sure MFA(in the transaction) and the transaction on the same node
+                %% don't need rpc again inside transaction.
+                case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
+                    {ok, Node} -> gen_server:call({?MODULE, Node}, MFA, Timeout);
+                    disconnected -> {error, disconnected}
+                end
+        end,
+    End = erlang:monotonic_time(),
+    Gap = erlang:convert_time_unit(Begin - End, native, millisecond) + 50,
+    RetryMs = get_retry_ms(),
+    confirm_commit(InitRes, SucceedNum, Gap, 3 * max(Gap, RetryMs)).
```

Comment:
Maybe calculate a `Deadline` before starting the wait loop, instead of calculating the next `Timeout` while looping.